### ISSUE TITLE
Avoid Counter-Productive Remediations (Melt-Down) during Zonal Outages

### ISSUE THREAD

**Issue by vlerenc** [gardener/machine-controller-manager#760]
**What would you like to be added**:
In case of an outage that leads to a network partitioning in which all nodes in one zone appear "offline" (while the cloud provider probably still shows the machines "online"), we start terminating (after `machineHealthTimeout`) the machines and removing the nodes one-by-one (thereby slowing down, but not avoiding the melt-down), because we can't/don't recognise the zone as network-partitioned. While this is a known fact, it's nonetheless harmful (at least not helpful), so can we do anything here to avoid that? See also #354.

Related, we drain the nodes (after `machineHealthTimeout`) parallel to the creation of the replacement machines, before the new node books in or becomes `Ready`. This can be a good thing (machine down, but sufficient extra capacity available), but in the case of a network partitioning, even draining is harmful (at least not helpful).

**Why is this needed**:
The disadvantages of the current behaviour seem to be:
- The outage is artificially prolonged as we lose machines/nodes and workload/pods and recreating those later after the cloud provider problem gets resolved takes additional time (rare machine types may not be available, which may complicate the situation further).
- If the outage lasts long enough and the last machine/node is gone, kube-scheduler will forget about the zone, which breaks topology spread constraints such as `DoNotSchedule` as it doesn't even remember that there is another zone it shall schedule a pod to (with `maxSkew` of 1, which we need for rolling updates, a second pod would be scheduled into a healthy zone instead of leaving the pod pending until the lost zone recovers).
- In large clusters, terminating tens of thousands of pods and recreating new ones may put additional stress on the control plane.

**Comment by elankath** [comment]
Went through #354 . IMHO, the primary problem is that in the MCM health-reconciliation we are slightly over-eager in marking a machine phase from `Running` ⮕  `Unknown` ⮕ (on health check time out) `Failed` leading to meltdown scenarios.  

Basically, if a Node is not ready, we immediately change the machine phase to `Unknown` followed by `Failed` on health check timeout.  The default health check timeout is 10m (when I last checked). IMHO is is too low. 

**Proposal**: If all the other "bad" Node conditions `KernelDeadlock,ReadonlyFilesystem,DiskPressure,NetworkUnavailable` are negative and the node `Ready` condition is false, then we should _explicitly_ check whether the VM is still present using `driver.getMachineStatus`. If the VM is fine and , then we should double the health check timeout before marking machine as failed.  IMHO, from my previous experience on cloud platforms, 10m is too low a timeout to mark a machine as failed due to network comm issues. 20m is more realistic.  However, we cannot generally increase the timeout - it should be done only specifically for the case when the Node is not Ready, there are no negative node conditions but the VM is present.

**Comment by himanshu-kun** [comment]
@vlerenc As per the meltdown fix that was introduced, we just remove one machine and wait for another one to join, if that doesn't happen we don't delete any other machine. Since kubelet is unable to reach the KAPI, so the machine obj we are waiting on never become `Running`.
We'll try simulating this locally and verify that its violated.

**Comment by vlerenc** [comment]
That was a misunderstanding @himanshu-kun on my part then. I mixed observation and assumption. I observed only 1 machine to die off, because my test clusters are tiny. I assumed that we would kill off every machine eventually, but if you wait for another machine to join, the process would stop. If that's the case, it's fine. Losing one machine is not too bad, because we cannot know for sure what's going on anyway.

@elankath 's proposal is also interesting. Then again, we discussed "best-guess" alternatives also in the past of similar nature, i.e. if all nodes in a zone become `Unknown`, it's fishy and any standard remediation is then probably doing more harm than good, but with small clusters that's difficult to assess. Then again, we could think it over and see whether we can get to some useful ruleset or whether the current behaviour is already sufficient.

@elankath 10m sounds short for you, but may be long in other cases. It's really difficult to balance that, if we have to balance it and cannot be smarter about the individual scenarios (because we don't know for sure).

**Comment by unmarshall** [comment]
Two additional points:

**Behavior of KCM (Node Controller):**
Node controller will wait for `--node-monitor-grace-period`(default: 40s, gardener-default: 2m) to mark the node as `Unknown` and then further wait for `--pod-eviction-timeout` (default: 5m) before it starts forcefully evicts pods from the nodes.
The rate at which pods are evicted is governed by `--node-eviction-rate` (see: https://kubernetes.io/docs/concepts/architecture/nodes/#rate-limits-on-eviction)
This means that while MCM attempts to preserve as many VMs as possible and slow down deletion of existing VMs, node controller will start to evict pods. If the outage lasts for hours, nodes will be removed, pods will be evicted but we will keep all but one machines per machine deployment running and thereby incurring cost.

**Further slowing down during meltdown**
Today we try and delete 1 machine per machine deployment. If a zone information is captured in every machine deployment across all worker pools defined by the customer, then it is technically possible to identify zone level issues by inspecting all such machine deployments. We can also then club all such machine deployments and do deletion of 1 machine from this larger set, slowing down the meltdown even further.

**Comment by vlerenc** [comment]
Before I make the same mistake once again, shall we close this ticket then, because it's actually wrong (what I write) and we are much better since the melt-down patch, because we wait for a new machine to come up, before we continue with the next machine.

As for KCM, that's indeed a follow-up issue that I saw as well. Then again, it's tough for us to make a decision, maybe pods should be evicted to come up elsewhere (if they allow that) and maybe that's counter-productive (because they won't come up elsewhere). We could try to be smart (mimic the scheduler), but it will probably be hard and not always correct? What we possibly could offer one day is giving people the information that a zone is "probably" down, not via text/Slack, but in a programmatic way. Then again, cloud provider issues are manifold (routers down, increased error rate, slow disk attach/detach, ... so many issues).

Anyway going back to the original issue, I will close this ticket now. If you feel otherwise, please reopen it again.
